src/main/java/org/broadinstitute/hellbender/engine/AssemblyRegion.java

```diff
@@ -0,0 +1,609 @@
+        final List<GATKRead> windowReads = new ArrayList<>();
+        for ( final GATKRead read : shard ) {
+            windowReads.add(read);
+        }
+
+        final LocusIteratorByState locusIterator = new LocusIteratorByState(windowReads.iterator(), DownsamplingMethod.NONE, false, false, ReadUtils.getSamplesFromHeader(readsHeader), readsHeader);
+        final ActivityProfile activityProfile = new BandPassActivityProfile(null, maxProbPropagationDistance, activeProbThreshold, BandPassActivityProfile.MAX_FILTER_SIZE, BandPassActivityProfile.DEFAULT_SIGMA, readsHeader);
+
+        // First, use our activity profile to determine the bounds of each assembly region:
+        List<AssemblyRegion> assemblyRegions = new ArrayList<>();
+        for ( final AlignmentContext pileup : locusIterator ) {
+            if ( ! activityProfile.isEmpty() ) {
+                final boolean forceConversion = pileup.getLocation().getStart() != activityProfile.getEnd() + 1;
+                assemblyRegions.addAll(activityProfile.popReadyAssemblyRegions(assemblyRegionPadding, minRegionSize, maxRegionSize, forceConversion));
+            }
+
+            if ( shard.getPaddedInterval().contains(pileup.getLocation()) ) {
+                final SimpleInterval pileupInterval = new SimpleInterval(pileup.getLocation());
+                final ReferenceBases refBase = new ReferenceBases(new byte[]{referenceContext.getBases()[pileup.getLocation().getStart() - referenceContext.getWindow().getStart()]}, pileupInterval);
+                final ReferenceContext pileupRefContext = new ReferenceContext(new ReferenceMemorySource(refBase, readsHeader.getSequenceDictionary()), pileupInterval);
+
+                final ActivityProfileState profile = evaluator.isActive(pileup, pileupRefContext, features);
+                activityProfile.add(profile);
+            }
+        }
+        assemblyRegions.addAll(activityProfile.popReadyAssemblyRegions(assemblyRegionPadding, minRegionSize, maxRegionSize, true));
+
+        // Then, fill the assembly regions with overlapping reads from the shard:
+        final PeekableIterator<GATKRead> reads = new PeekableIterator<>(windowReads.iterator());
+        AssemblyRegion lastRegion = null;
+        for ( final AssemblyRegion region : assemblyRegions ) {
+            if ( lastRegion != null ) {
+                for ( final GATKRead lastRegionRead : lastRegion.getReads() ) {
+                    if ( region.getExtendedSpan().overlaps(lastRegionRead) ) {
+                        region.add(lastRegionRead);
+                    }
+                }
+            }
+
+            while ( reads.hasNext() ) {
+                final GATKRead read = reads.peek();
+                if ( region.getExtendedSpan().overlaps(read) ) {
+                    region.add(reads.next());
+                }
+                else {
+                    break;
```

Comment:
why is there a break here - is it because there's a sorting assumption? If yes, make this assumption explicit